File: Leetcode/1472 Design Browser History.cpp
class BrowserHistory {
    struct Node {
        string url;
        Node *prev, *next;
        
        Node(const string &url, Node *prev, Node *next) 
            : url(url), prev(prev), next(next) {}
        
        Node(const string &url) 
            : url(url), prev(nullptr), next(nullptr) {}
    };
    
    void delete_list(Node *node) {
        while(node != nullptr) {
            Node *to_delete = node;
            node = node->next;
            delete to_delete;
        }
    }
    
    Node *current;
public:
    BrowserHistory(string homepage) {
        current = new Node(homepage);
    }
    
    void visit(string url) {
        delete_list(current->next);
        current = current->next = new Node(url, current, nullptr);
    }
    
    string back(int steps) {
        int n_steps = 0;
        while(current->prev != nullptr and n_steps < steps) {
            current = current->prev;
            ++n_steps;
        }
        return current->url;
    }
    
    string forward(int steps) {
        int n_steps = 0;
        while(current->next != nullptr and n_steps < steps) {
            current = current->next;     
            ++n_steps;
        }
        return current->url;
    }
};

File: Leetcode/1472 Design Browser History.ts
type Optional<T> = T | null;

class BrowserHistoryNode {
    url: string;
    prev: Optional<BrowserHistoryNode>;
    next: Optional<BrowserHistoryNode>;

    constructor(url: string, prev: Optional<BrowserHistoryNode>, next: Optional<BrowserHistoryNode>) {
        this.url = url;
        this.prev = prev;
        this.next = next;
    }
}

class BrowserHistory {
    curr: Optional<BrowserHistoryNode>;

    constructor(homepage: string) {
        this.curr = new BrowserHistoryNode(homepage, null, null);
    }

    visit(url: string): void {
        this.curr = this.curr.next = new BrowserHistoryNode(url, this.curr, null);
    }

    back(steps: number): string {
        let nSteps = 0;
        while(this.curr.prev != null && nSteps < steps) {
            this.curr = this.curr.prev;
            nSteps++;
        }
        return this.curr.url;
    }

    forward(steps: number): string {
        let nSteps = 0;
        while(this.curr.next != null && nSteps < steps) {
            this.curr = this.curr.next;
            nSteps++;
        }
        return this.curr.url;
    }
}
